GLMakie/src/drawing_primitives.jl

```diff
@@ -819,33 +819,34 @@ function draw_atomic(screen::Screen, scene::Scene, plot::Surface)
         if all(T -> T <: Union{AbstractMatrix, AbstractVector}, types)
             t = Makie.transform_func_obs(plot)
             mat = plot[3]
-            xypos = lift(plot, pop!(gl_attributes, :f32c), plot.model, t, plot[1], plot[2], space) do f32c, model, t, x, y, space
+            xyzpos = lift(plot, pop!(gl_attributes, :f32c), plot.model, t, plot[1], plot[2], mat, space) do f32c, model, t, x, y, z, space
                 # Only if transform doesn't do anything, we can stay linear in 1/2D
                 if Makie.is_identity_transform(t) && isnothing(f32c)
-                    return (x, y)
+                    return (x, y, z)
                 elseif Makie.is_translation_scale_matrix(model)
                     matrix = if x isa AbstractMatrix && y isa AbstractMatrix
-                        Makie.f32_convert(f32c, apply_transform.((t,), Point.(x, y), space), space)
+                        [Makie.f32_convert(f32c, apply_transform(t, Point3(x, y, z[ix, iy]), space), space) for (ix, x) in pairs(x), (iy, y) in pairs(y)]
```

Comment:
Isn't this calling `z[CartesianIndex(...), CartesianIndex(...)]`?

Reply:
Should probably extend the refimg test to test 

| vec-vec | vec-mat |
| --- | --- |
| mat-vec | mat-mat |

inputs

Reply:
Good point, and will test that.

Reply:
This could probably also be a loop operating on a pre-allocated array - that would be much clearer I think.